### PR TITLE
[feat] start game use case

### PR DESCRIPTION
**Pull request by RadicalV**

### CODE REVIEW

File: server/battleships/gateway/implementation/src/main/java/eu/optas/gateway/implementation/InMemoryGameGateway.java [hunk 1]
```diff
@@ -1,15 +1,55 @@
...
+        Game game = new Game(id, GameState.IN_PROGRESS, new Board(grid, ships), 25, 0);
+        gameList.add(game);
+        return game;
+    }
+
+    private String generateId() {
+        return UUID.randomUUID().toString();
+    }
+
+    private List<List<Integer>> generateGrid() {
```
Comment (DovydasSp) [inline server/battleships/gateway/implementation/src/main/java/eu/optas/gateway/implementation/InMemoryGameGateway.java:36]: Pradžioje gali pasikurti kažkokį failą, kur susidedi kelis pregenerated grid'us su laivais ir random vieną iš jų paimi žaidimui. Vėliau jei bus laiko, galėsi ir generavimo logiką pasirašyti

File: server/battleships/gateway/implementation/src/main/java/eu/optas/gateway/implementation/InMemoryGameGateway.java [hunk 2]
```diff
@@ -1,15 +1,55 @@
...
+        List<List<Integer>> grid = new ArrayList<>();
+        for (int i = 0; i < 10; i++) {
+            List<Integer> row = new ArrayList<>();
+            for (int j = 0; j < 10; j++) {
```
Comment (martynas-openg) [inline server/battleships/gateway/implementation/src/main/java/eu/optas/gateway/implementation/InMemoryGameGateway.java]: toks menkniekis, bet mes trinam bracketus jeigu tik vienas statementas bloke

File: server/battleships/gateway/implementation/src/test/java/eu/optas/gateway/implementation/InMemoryGameGatewayTest.java [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+import org.junit.jupiter.api.Test;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+class InMemoryGameGatewayTest {
+    InMemoryGameGateway inMemoryGameGateway;
```
Comment (martynas-openg) [inline server/battleships/gateway/implementation/src/test/java/eu/optas/gateway/implementation/InMemoryGameGatewayTest.java]: private

File: server/battleships/gateway/implementation/src/main/java/eu/optas/gateway/implementation/InMemoryGameGateway.java [hunk 4]
```diff
@@ -1,15 +1,55 @@
...
+import eu.optas.utils.Coordinates;
+import eu.optas.utils.GameState;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.List;
+import java.util.UUID;
 
 public class InMemoryGameGateway implements GameGateway {
     private List<Game> gameList;
```
Comment (martynas-openg) [inline server/battleships/gateway/implementation/src/main/java/eu/optas/gateway/implementation/InMemoryGameGateway.java]: private final

File: server/battleships/gateway/implementation/src/test/java/eu/optas/gateway/implementation/InMemoryGameGatewayTest.java [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+    @BeforeEach
+    void setUp() {
+        inMemoryGameGateway = new InMemoryGameGateway();
+    }
+
+    @Test
+    void createGame() {
+        Game game = inMemoryGameGateway.createGame();
+        List<List<Integer>> expectedGrid = new ArrayList<>(Collections.nCopies(10, new ArrayList<>(Collections.nCopies(10, 0))));
```
Comment (martynas-openg) [inline server/battleships/gateway/implementation/src/test/java/eu/optas/gateway/implementation/InMemoryGameGatewayTest.java:25]: blem kaip tas List<List<...>> matrica cancer atrodo

File: server/battleships/use_cases/implementation/src/test/java/eu/optas/use_cases/implementation/GameD2BConverterTest.java [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+                "123",
+                GameState.IN_PROGRESS,
+                board,
+                25,
+                0
+        );
+        BoundaryGame expectedGame = new BoundaryGame(
+                "123",
+                GameState.IN_PROGRESS,
+                boardD2BConverter.convert(board),
```
Comment (DovydasSp) [inline server/battleships/use_cases/implementation/src/test/java/eu/optas/use_cases/implementation/GameD2BConverterTest.java]: kažkaip keistai čia

File: server/battleships/main/src/main/java/eu/optas/Main.java [hunk 7]
```diff
@@ -14,11 +17,15 @@ public class Main {
...
+        ShipD2BConverter shipD2BConverter = new ShipD2BConverter();
+        BoardD2BConverter boardD2BConverter = new BoardD2BConverter(shipD2BConverter);
+        GameD2BConverter gameD2BConverter = new GameD2BConverter(boardD2BConverter);
+
+        StartGameUC startGameUC = new StartGameInteractor(gameGateway, gameD2BConverter);
 
         Javalin app = Javalin.create().start(3000);
 
-        app.get("/startGame", startGameRoute::startGame);
+        app.post("/games", new StartGameRoute(startGameUC));
```
Comment (martynas-openg) [inline server/battleships/main/src/main/java/eu/optas/Main.java]: nice, rest

File: server/battleships/rest/src/main/java/eu/optas/rest/StartGameRoute.java [hunk 8]
```diff
@@ -2,15 +2,19 @@
...
 
     public StartGameRoute(StartGameUC startGameUC) {
         this.startGameUC = startGameUC;
     }
 
-    public void startGame(Context context) {
-        context.result("Start Game Route!");
+    @Override
+    public void handle(@NotNull Context ctx) {
+        ctx.json(startGameUC.startGame());
```
Comment (martynas-openg) [inline server/battleships/rest/src/main/java/eu/optas/rest/StartGameRoute.java]: json yra rest (View) layerio dalykai, reiktų boundary pakonvertuoti į rest prieš dedant į response.
Čia kadangi strict clean architecture darai.

File: server/battleships/rest/src/test/java/eu/optas/rest/StartGameRouteTest.java [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+        when(startGameUC.startGame()).thenReturn(returnedGame);
+
+        StartGameRoute startGameRoute = new StartGameRoute(startGameUC);
+        Context ctx = mock(Context.class);
+
+        startGameRoute.handle(ctx);
+
+        verify(ctx).json(returnedGame);
+        verify(ctx).status(201);
+        verify(startGameUC, times(1)).startGame();
```
Comment (martynas-openg) [inline server/battleships/rest/src/test/java/eu/optas/rest/StartGameRouteTest.java]: jeigu times(1), tai gali rašyti tiesiog verify(startGameUC).startGame();

File: server/battleships/use_cases/implementation/src/test/java/eu/optas/use_cases/implementation/GameD2BConverterTest.java (hunk 6 above)
Comment (martynas-openg) [inline server/battleships/use_cases/implementation/src/test/java/eu/optas/use_cases/implementation/GameD2BConverterTest.java]: šitas statement grąžina null, tai realiai implementacijoj tą vietą vietoj converterio kvietimo galėtum pakeist į null ir testai praeitų. Reiktų užsimockinti kažkokią reikšmę boardConverterio rezultatui ir paasertinti, kad tikrai game outpute yra expectinama board reikšmė gauta iš vidinio konverterio

Comment (martynas-openg) [reply server/battleships/use_cases/implementation/src/test/java/eu/optas/use_cases/implementation/GameD2BConverterTest.java]: Tas pats dalykas ir BoardD2BConvertery matau

File: server/battleships/use_cases/implementation/src/test/java/eu/optas/use_cases/implementation/GameD2BConverterTest.java [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+                25,
+                0
+        );
+
+        BoundaryGame convertedGame = gameD2BConverter.convert(inputGame);
+
+        assertThat(convertedGame.getId()).isEqualTo(expectedGame.getId());
```
Comment (martynas-openg) [inline server/battleships/use_cases/implementation/src/test/java/eu/optas/use_cases/implementation/GameD2BConverterTest.java]: labai geras metodas yra assertThat(...).usingRecursiveComparison().isEqualTo(...), tai tau objektus sulygins pagal visus jų fieldus, o priedu gali po usingRecursiveComparison įdėti ignoringFields, arba paspecifyint fieldams comparatorius.

File: server/battleships/use_cases/implementation/src/main/java/eu/optas/use_cases/implementation/GameD2BConverter.java [hunk 11]
```diff
@@ -0,0 +1,24 @@
+package eu.optas.use_cases.implementation;
+
+import eu.optas.domain.Game;
+import eu.optas.use_cases.api.BoundaryGame;
+import eu.optas.utils.Converter;
+
+public class GameD2BConverter extends Converter<Game, BoundaryGame> {
+    private BoardD2BConverter boardD2BConverter;
```
Comment (martynas-openg) [inline server/battleships/use_cases/implementation/src/main/java/eu/optas/use_cases/implementation/GameD2BConverter.java]: private final

File: server/battleships/utils/src/main/java/eu/optas/utils/Converter.java [hunk 12]
```diff
@@ -0,0 +1,16 @@
+package eu.optas.utils;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public abstract class Converter<I, O> {
+    public List<O> convertAll(List<I> input) {
+        List<O> output = new ArrayList<>();
```
Comment (martynas-openg) [inline server/battleships/utils/src/main/java/eu/optas/utils/Converter.java]: gali iškart basics mokintis javos Stream API.
Šitą vietą galima būtų perdaryti taip:
return input.stream()
                  .map(this::convert)
                  .collect(Collectors.toList());
                  
Čia labai panašiai kaip ir javascripte su collectionais dirbi

File: server/battleships/use_cases/implementation/src/test/java/eu/optas/use_cases/implementation/BoardD2BConverterTest.java [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+
+class BoardD2BConverterTest {
+
+    @Test
+    void convert() {
+        ShipD2BConverter shipD2BConverterMock = mock(ShipD2BConverter.class);
+        BoardD2BConverter boardD2BConverter = new BoardD2BConverter(shipD2BConverterMock);
+
+        List<List<Integer>> grid = new ArrayList<>();
+        grid.add(new ArrayList<>(Arrays.asList(0, 0, 0, 0)));
```
Comment (martynas-openg) [inline server/battleships/use_cases/implementation/src/test/java/eu/optas/use_cases/implementation/BoardD2BConverterTest.java]: new ArrayList<>() gali versti į List.of() (tik žinok, kad tas gautas list bus unmodifiable, tai jeigu bandysi modifikuoti - mes exception)

File: server/battleships/gateway/implementation/src/main/java/eu/optas/gateway/implementation/InMemoryGameGateway.java (hunk 1 above)
Comment (RadicalV) [reply server/battleships/gateway/implementation/src/main/java/eu/optas/gateway/implementation/InMemoryGameGateway.java:36]: kolkas paliksiu taip, vėliau žiurėsiu

File: server/battleships/use_cases/implementation/src/test/java/eu/optas/use_cases/implementation/GameD2BConverterTest.java (hunk 6 above)
Comment (RadicalV) [reply server/battleships/use_cases/implementation/src/test/java/eu/optas/use_cases/implementation/GameD2BConverterTest.java]: IDE padaro ta labai dideli indent

File: server/battleships/rest/src/main/java/eu/optas/rest/Boundary2RestConverter.java [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+
+import eu.optas.use_cases.api.BoundaryBoard;
+import eu.optas.use_cases.api.BoundaryGame;
+import eu.optas.use_cases.api.BoundaryShip;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class Boundary2RestConverter {
+    public Boundary2RestConverter() {
```
Comment (martynas-openg) [inline server/battleships/rest/src/main/java/eu/optas/rest/Boundary2RestConverter.java]: tuščią konstruktorių gali ištrinti, nes jis by default yra visose klasėse (nematomas)

File: server/battleships/rest/src/test/java/eu/optas/rest/Boundary2RestConverterTest.java [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+    }
+
+    @Test
+    void convertBoard() {
+        BoundaryShip boundaryShip = mock(BoundaryShip.class);
+        RestShip restShip = mock(RestShip.class);
+        BoundaryBoard boundaryBoard = new BoundaryBoard(List.of(List.of(0, 0, 0)), List.of(boundaryShip));
+        RestBoard expectedBoard = new RestBoard(List.of(List.of(0, 0, 0)), List.of(restShip));
+
+        RestBoard convertedBoard = boundary2RestConverter.convertBoard(boundaryBoard);
```
Comment (martynas-openg) [inline server/battleships/rest/src/test/java/eu/optas/rest/Boundary2RestConverterTest.java]: ne taip supratai, viską gali patestuoti per vieną public metodą "convertGame"

Comment (martynas-openg) [reply server/battleships/rest/src/test/java/eu/optas/rest/Boundary2RestConverterTest.java]: ir užteks tiesiog vienam teste šitą konvertavimą patestuot, nes nėra pas tave čia kažkokiu svarbių edge case konvertavimo logikoj

File: server/battleships/use_cases/implementation/src/test/java/eu/optas/use_cases/implementation/GameD2BConverterTest.java [hunk 16]
```diff
@@ -27,17 +32,12 @@ void convert() {
...
         );
-
         BoundaryGame convertedGame = gameD2BConverter.convert(inputGame);
 
-        assertThat(convertedGame.getId()).isEqualTo(expectedGame.getId());
-        assertThat(convertedGame.getState()).isEqualTo(expectedGame.getState());
-        assertThat(convertedGame.getBoard()).isEqualTo(expectedGame.getBoard());
-        assertThat(convertedGame.getHitsRemaining()).isEqualTo(expectedGame.getHitsRemaining());
-        assertThat(convertedGame.getShipsDestroyed()).isEqualTo(expectedGame.getShipsDestroyed());
+        assertThat(convertedGame).usingRecursiveComparison().isEqualTo(expectedGame);
```
Comment (martynas-openg) [inline server/battleships/use_cases/implementation/src/test/java/eu/optas/use_cases/implementation/GameD2BConverterTest.java:41]: šitas testas paeina net kai ir užmockintas board?

Comment (RadicalV) [reply server/battleships/use_cases/implementation/src/test/java/eu/optas/use_cases/implementation/GameD2BConverterTest.java:41]: yep

Comment (martynas-openg) [reply server/battleships/use_cases/implementation/src/test/java/eu/optas/use_cases/implementation/GameD2BConverterTest.java:41]: best practice būtų atskirai tą board paasertinti šitam teste (kad grįžo tas specifinis instance iš mockinto konverterio, o ne objektas su defaultiniais fieldais kuris rekursyviai lygus mockui)

**Comment by DovydasSp** [comment]
^ čia ne commit message :D

**Review by martynas-openg — APPROVED** [review]

**Review by DovydasSp — APPROVED** [review]

**Review by WeebDetector — APPROVED** [review]
